fsimpl/unionfs/unionfs.go

```diff
@@ -0,0 +1,171 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionfs
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/hugelgupf/p9/fsimpl/readdir"
+	"github.com/hugelgupf/p9/fsimpl/templatefs"
+	"github.com/hugelgupf/p9/linux"
+	"github.com/hugelgupf/p9/p9"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	ErrFlatHierarchy = errors.New("unionfs only supports a flat hierarchy")
+	ErrFileExists    = errors.New("file already exists")
+)
+
+type Opt func(fs *FS) error
+
+// FS is a p9.Attacher.
+type FS struct {
+	mounts map[string]p9.File
+}
+
+func WithMount(dir string, attacher p9.Attacher) Opt {
+	return func(fs *FS) error {
+		if strings.Contains(dir, string(filepath.Separator)) {
+			return fmt.Errorf("%w: directories are not supported: %s", ErrFlatHierarchy, dir)
+		}
+		if _, ok := fs.mounts[dir]; ok {
+			return fmt.Errorf("%w: %s", ErrFileExists, dir)
+		}
+
+		f, err := attacher.Attach()
+		if err != nil {
+			return err
+		}
+		fs.mounts[dir] = f
+		return nil
+	}
+}
+
+func WithDir(dir string, mounts ...Opt) Opt {
+	return func(fs *FS) error {
+		subfs, err := New(mounts...)
+		if err != nil {
+			return err
+		}
+
+		return WithMount(dir, subfs)(fs)
+	}
+}
+
+func WithFile(file string, f p9.File) Opt {
+	return func(fs *FS) error {
+		if strings.Contains(file, "/") {
+			return fmt.Errorf("%w: directories are not supported: %s", ErrFlatHierarchy, file)
+		}
+		if _, ok := fs.mounts[file]; ok {
+			return fmt.Errorf("%w: %s", ErrFileExists, file)
+		}
+
+		fs.mounts[file] = f
+		return nil
+	}
+}
+
+func New(mounts ...Opt) (*FS, error) {
+	fs := &FS{
+		mounts: make(map[string]p9.File),
+	}
+	for _, m := range mounts {
+		if err := m(fs); err != nil {
+			return nil, err
+		}
+	}
+	return fs, nil
+}
+
+type root struct {
+	p9.DefaultWalkGetAttr
+	templatefs.ReadOnlyDir
+	templatefs.NilCloser
+
+	fs *FS
+}
+
+// Attach implements p9.Attacher.Attach.
+func (fs *FS) Attach() (p9.File, error) {
+	return &root{fs: fs}, nil
+}
+
+var (
+	_ p9.File     = &root{}
+	_ p9.Attacher = &FS{}
+)
+
+var (
+	rootQID = p9.QID{Type: p9.TypeDir, Path: 1, Version: 0}
+)
+
+// Walk implements p9.File.Walk.
+func (r *root) Walk(names []string) ([]p9.QID, p9.File, error) {
+	if len(names) == 0 {
+		return []p9.QID{rootQID}, &root{fs: r.fs}, nil
+	}
+
+	file, ok := r.fs.mounts[names[0]]
+	if !ok {
+		return nil, nil, linux.ENOENT
+	}
+	// Let the file figure out its own QID.
+	return file.Walk(names[1:])
```

Comment:
yes, QIDS are allowed to change, walk to walk, since just about anything can be happening on the server. So no problem there.

Reply:
One QID problem I've been thinking of - different file systems underneath can return the same QID for different file paths. Is that an issue? I need to read the documentation again and see how the Linux clients handles them.

Reply:
Made a thing for that.